### PR TITLE
Speed up build time by optimizing test runs

### PR DESCRIPTION
**Pull request by a-sayyed**
This PR attempts to speed up the project build time by tackling the different issues that result in a slow build
- All tests run with the `CleanupTestExecutionListener` which cleans up all data created as part of the test run for the `DEFAULT` tenant
- All tests run in the context of the same tenant (`DEFAULT`), which makes it very difficult to even consider parallelizing the test runs
- The `@DirtiesContext` introduced in the `AbstractIntegrationTest` means that after each test class, the spring context is destroyed and recreated (including the database migration scripts)
- The 'quota' tests take much longer than other tests because they always have to reach the quota limits.

Goals:
- Introducing shared contexts https://stackoverflow.com/a/23004557
- All tests methods should run with their own tenant, this means we don't have to do any clean up, and, potentially, we could parallelize the tests since each run is completely isolated from the other.
- ~~Grouping related (by spring configuration) tests together into suits, this would ensure that at least the tests within a suite would run in the same spring context - and, if needed, the 'dirty' context could be disposed of before the next test suite run.~~ This is not needed if we use Context Caching
- If possible, completely get rid of the `@DirtiesContext` annotation from all test runs
- Reducing quota limits, so the tests will hit them faster
- Not randomly creating the database, since the migration takes a while

### CODE REVIEW

File: hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/util/WithSqlDatabaseRule.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        executeStatement(username, password, uri.split("/" + schemaName)[0], schemaName,
+                "CREATE SCHEMA IF NOT EXISTS " + schemaName + ";");
+        databaseName = schemaName;
+    }
+
+    private static boolean isRunningWithMsSql() {
+        return "SQL_SERVER".equals(System.getProperty("spring.jpa.database"));
+    }
+
+    public static void createMsSqlSchemaIfNotExists(final String username, final String password, final String uri) {
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/util/WithSqlDatabaseRule.java:69]: todo: test with mssql

File: hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/util/AbstractIntegrationTest.java [hunk 2]
```diff
@@ -373,35 +366,35 @@ protected Action prepareFinishedUpdate(final String controllerId, final String d
...
+
+            runtimeType = securityRule.runAsPrivileged(
+                    () -> testdataFactory.findOrCreateSoftwareModuleType(TestdataFactory.SM_TYPE_RT));
+            runtimeType = securityRule.runAsPrivileged(() -> softwareModuleTypeManagement.update(
+                    entityFactory.softwareModuleType().update(runtimeType.getId()).description(description)));
+
+            standardDsType = securityRule.runAsPrivileged(() -> testdataFactory.findOrCreateDefaultTestDsType());
+        } catch (final Exception e) {
+            LOG.error("Not creating default software types since tenant {} has autoCreate = false",
+                    tenantAware.getCurrentTenant(), e);
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/util/AbstractIntegrationTest.java:396]: don't log the error as error, better as trace

File: hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/TestConfiguration.java [hunk 3]
```diff
@@ -137,19 +141,28 @@ DownloadIdCache downloadIdCache() {
...
-                applicationEventFilter);
+    SimpleApplicationEventMulticaster applicationEventMulticaster() {
+        final SimpleApplicationEventMulticaster simpleApplicationEventMulticaster = new FilterEnabledApplicationEventPublisher();
         simpleApplicationEventMulticaster.setTaskExecutor(asyncExecutor());
+        simpleApplicationEventMulticaster.addApplicationListener(tenantEventCounter());
         return simpleApplicationEventMulticaster;
     }
 
-    private static class FilterEnabledApplicationEventPublisher extends SimpleApplicationEventMulticaster {
+    public static class FilterEnabledApplicationEventPublisher extends SimpleApplicationEventMulticaster {
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/TestConfiguration.java:159]: is this still needed?

File: hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/matcher/EventVerifier.java [hunk 4]
```diff
@@ -9,179 +9,143 @@
...
-    public static void publishResetMarkerEvent(final ApplicationEventPublisher publisher) {
-        publisher.publishEvent(new ResetCounterMarkerEvent());
-    }
-
-    @Override
-    public void beforeTestMethod(final TestContext testContext) throws Exception {
-        final Optional<Expect[]> expectedEvents = getExpectationsFrom(testContext.getTestMethod());
-        expectedEvents.ifPresent(events -> beforeTest(testContext));
-    }
+//    private final TenantEventCounter eventCounter = new TenantEventCounter();
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-test/src/main/java/org/eclipse/hawkbit/repository/test/matcher/EventVerifier.java:40]: delete

File: hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/AmqpProblematicMessageHandlerServiceIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    @After
+    public void resetControllerManagement() {
+        amqpMessageHandlerService.setControllerManagement(controllerManagement);
+    }
+
+    @Test
+    @Description("Messages that result into certain exceptions being raised should not be requeued. This message should forwarded to the deadletter queue")
+    @ExpectEvents({@Expect(type = TargetCreatedEvent.class, count = 0)})
+    public void ignoredExceptionTypesShouldNotBeRequeued() {
```
Comment (bogdan-bondar) [inline hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/AmqpProblematicMessageHandlerServiceIntegrationTest.java:63]: Could you please explain the reason behind extracting this method to its own class? What context changes are required and why do we need a fresh context for it?

File: hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/listener/DeadletterListener.java [hunk 6]
```diff
@@ -8,18 +8,27 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
 import org.eclipse.hawkbit.rabbitmq.test.listener.TestRabbitListener;
 import org.springframework.amqp.core.Message;
 import org.springframework.amqp.rabbit.annotation.RabbitListener;
 
 public class DeadletterListener implements TestRabbitListener {
 
-    public static final String LISTENER_ID = "deadletter";
+    private final AtomicInteger messageCount = new AtomicInteger(0);
```
Comment (bogdan-bondar) [inline hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/listener/DeadletterListener.java:19]: Does it make sense to introduce multi tenancy here?

File: hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/listener/ReplyToListener.java [hunk 7]
```diff
@@ -29,77 +28,89 @@
     public static final String LISTENER_ID = "replyto";
     public static final String REPLY_TO_QUEUE = "reply_queue";
 
-    private final Map<EventTopic, List<Message>> eventMessages = new EnumMap<>(EventTopic.class);
-    private final List<EventTopic> eventMessageTopics = new ArrayList<>();
-    private final Map<String, Message> deleteMessages = new HashMap<>();
-    private final Map<String, Message> pingResponseMessages = new HashMap<>();
+    private final Map<String, Map<EventTopic, List<Message>>> tenantEventMessages = new ConcurrentHashMap<>();
```
Comment (bogdan-bondar) [inline hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/listener/ReplyToListener.java:31]: I guess you can hardcode LISTENER_ID

**Comment by a-sayyed** [comment]
Will introduce multiple smaller PRs instead

**Pull request closed without merging**